File: day5/day5.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func findRow(line string) int {
	binaryString := strings.Replace(strings.Replace(line, "B", "1", 7), "F", "0", 7)
	row, _ := strconv.ParseInt(binaryString, 2, 7)
	return int(row)
}

func findCol(line string) int {
	binaryString := strings.Replace(strings.Replace(line, "R", "1", 3), "L", "0", 3)
	col, _ := strconv.ParseInt(binaryString, 2, 3)
	return int(col)
}

func calculateSeatID(row int, col int) int {
	return row*8 + col
}

func main() {
	highestSeatID := 0
	maxSeatID := 127*8 + 7
	occupiedSeats := make([]int, maxSeatID)

	file, _ := os.Open("day5.in")
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		row := findRow(line[:7])
		col := findCol(line[7:])
		seatID := calculateSeatID(row, col)
		occupiedSeats[seatID] = seatID
		if seatID > highestSeatID {
			highestSeatID = seatID
		}
	}

	fmt.Println("Highest seat ID:", highestSeatID)
	for i := 1; i < len(occupiedSeats[:highestSeatID+1]); i++ {
		if occupiedSeats[i] == 0 && occupiedSeats[i-1] > 0 && occupiedSeats[i+1] > 0 {
			fmt.Println("My seat ID:", i)
		}
	}
}

File: day5/day5.py
def find_row(line):
    binary_string = line.replace('B', '1').replace('F', '0')
    return int(binary_string, 2)

def find_col(line):
    binary_string = line.replace('R', '1').replace('L', '0')
    return int(binary_string, 2)

def calculate_seat_ID(row, col):
    return row * 8 + col

def main():
    highest_seat_ID = 0
    max_seat_id = 127*8+7
    occupied_seats = [0] * max_seat_id
    with open("day5.in") as file:
        for line in file:
            row = find_row(line[:7])
            col = find_col(line[7:])
            seat_ID = calculate_seat_ID(row, col)
            occupied_seats[seat_ID] = seat_ID
            if seat_ID > highest_seat_ID:
                highest_seat_ID = seat_ID
    print("Highest seat ID:",highest_seat_ID)
    for i in range(1, len(occupied_seats[:highest_seat_ID+1])):
        if occupied_seats[i] == 0 and occupied_seats[i-1] > 0 and occupied_seats[i+1] > 0:
            print("My seat ID:", i)


main()